pkg/managed/service.go

```diff
@@ -114,118 +113,121 @@ func (m *Service) GetLabels(nodeGroupName string) (map[string]string, error) {
-	if err != nil {
-		return err
+	if kubernetesVersion != "" {
+		if _, err := semver.ParseTolerant(kubernetesVersion); err != nil {
+			return errors.Wrap(err, "error parsing Kubernetes version")
+		}
+		updateInput.Version = aws.String(kubernetesVersion)
+	} else {
+		updateInput.Version = nodegroupOutput.Nodegroup.Version
 	}
 
-	ssmOutput, err := m.provider.SSM().GetParameter(&ssm.GetParameterInput{
-		Name: &ssmParameterName,
-	})
+	nodegroupUpdate, err := m.provider.EKS().UpdateNodegroupVersion(updateInput)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "error upgrading nodegroup")
 	}
 
-	imageID := *ssmOutput.Parameter.Value
-
-	// To get the Kubernetes patch version, as it's not reported in the SSM GetParameter call
-	imagesOutput, err := m.provider.EC2().DescribeImages(&ec2.DescribeImagesInput{
-		ImageIds: aws.StringSlice([]string{imageID}),
-	})
-
-	if err != nil {
-		return err
-	}
-
-	if len(imagesOutput.Images) != 1 {
-		return fmt.Errorf("expected to find exactly 1 image; got %d", len(imagesOutput.Images))
+	if updateErrors := nodegroupUpdate.Update.Errors; len(updateErrors) > 0 {
+		return errors.Errorf("failed to upgrade nodegroup:\n%v", aggregateErrors(updateErrors))
 	}
 
-	image := *imagesOutput.Images[0]
-	amiReleaseVersion, err := extractAMIReleaseVersion(*image.Name)
-	if err != nil {
-		return errors.Wrap(err, "error extracting AMI release version")
+	for _, param := range nodegroupUpdate.Update.Params {
+		if *param.Type == eks.UpdateParamTypeReleaseVersion {
+			newReleaseVersion := *param.Value
+			if newReleaseVersion == *nodegroupOutput.Nodegroup.ReleaseVersion {
+				logger.Info("nodegroup %q is already up-to-date (release version: %v)", nodeGroupName, *nodegroupOutput.Nodegroup.ReleaseVersion)
+				return nil
+			}
+			logger.Info("upgrading nodegroup to release version %v", newReleaseVersion)
+		}
 	}
 
-	kubernetesVersion, err = extractKubeVersion(*image.Description)
-	if err != nil {
-		return errors.Wrap(err, "error extracting Kubernetes version")
+	if waitTimeout > 0 {
+		ctx, cancelFunc := context.WithTimeout(context.Background(), waitTimeout)
+		defer cancelFunc()
+		return m.waitForUpdate(ctx, nodeGroupName, nodegroupUpdate.Update.Id)
 	}
-	releaseVersion := makeReleaseVersion(kubernetesVersion, amiReleaseVersion)
-	if releaseVersion == *nodeGroup.ReleaseVersion {
-		logger.Info("nodegroup %q is already up-to-date", nodeGroupName)
-		return nil
-	}
-	return m.updateNodeGroupVersion(nodeGroupName, releaseVersion)
+	logger.Info("nodegroup upgrade request submitted")
+	return nil
 }
 
-func (m *Service) updateNodeGroupVersion(nodeGroupName, releaseVersion string) error {
-	template, err := m.stackCollection.GetManagedNodeGroupTemplate(nodeGroupName)
-	if err != nil {
-		return err
-	}
+func (m *Service) waitForUpdate(ctx context.Context, nodeGroupName string, updateID *string) error {
```

Comment:
can this code be replaced with the function here: https://github.com/weaveworks/eksctl/blob/8faceb2d3f5ae1b1065d24624615866eac16e7da/pkg/eks/update.go#L201

Reply:
I avoided using it because that function is tied to `ClusterProvider` and uses fields, like the `WaitTimeout`, that are set when `ClusterProvider` is created and can't be changed later. We could refactor it but that'd require changing all calling code, which I think would've been out of scope of this PR, but I think we can address it in a separate PR. Moreover, the `DescribeUpdateRequest` in `waitForUpdateToSucceed` lacks the `NodegroupName` field, so I'm not sure if it'd still work the same (although I think the update ID should suffice).